flow/ProtocolVersion.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * ProtocolVersion.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <cstdint>
+
+namespace fdb_versions {
+
+// Whenever a new breaking feature is introduced in FDB for which we need code
+// that tests whether the serialized version of the code was serialized with
+// that feature, the author should add this macro with the name of the feature
+// to the corresponding version. Users can than later call `version->hasFeature()`
+// on a deserialized `ProtocolVersion` to test for the feature.
+#define VERSION_FEATURE(x)                                                                                             \
+	constexpr bool has##x() const { return static_cast<const P*>(this)->version() >= protocolVersion; }
+
+// A version-class. These need to be ordered by version and the ordering
+// happens through inheritance. Each of these structs need to know about
+// the `ProtocolVersion`-class (through the template parameter) and it has
+// to be names `P` (otherwise the VERSION_FEATURE macro won't work anymore).
+// These classes are used to determine if a feature was available at a certain
+// protocol version. It does determine that through static polymorphism.
+template<class P>
+struct v5_5 {
+	static constexpr uint64_t protocolVersion = 0x0FDB00A551000000LL;
+
+	VERSION_FEATURE(MultiVersionClient)
+};
+
+template<class P>
+struct v5_6 : v5_5<P> {
+	static constexpr uint64_t protocolVersion = 0x0FDB00A560010001LL;
+
+	VERSION_FEATURE(TagLocality)
+};
+
+template <class P>
+struct v6_0 : v5_6<P> {
+	static constexpr uint64_t protocolVersion = 0x0FDB00B060000001LL;
+
+	VERSION_FEATURE(Fearless)
+};
+
+template <class P>
+struct v6_1 : v6_0<P> {};
```

Comment:
A few thoughts about these version structs:

Is the idea that we'd have (at least) one per release version, or just one per protocol version that introduces any new features? If we go with the latter, does the empty `v6_1` struct serve any purpose?

The names of these structs look roughly like our release versions, though I think that may be introducing some confusion here. We don't have a 5.5 or 5.6 (these versions roughly correspond with 5.1 and 6.0-PRE, I think), and in all of these cases the protocol version doesn't actually line up with the protocol version of the release (see list below). Which leads me to another issue, which is the potential need to specify protocol versions that arise between releases. It may not actually be intentional that these numbers look like our release version numbers, but even so I think the chance for confusion exists.

I think we should decide whether these feature checks need to be associated with release versions, and if they do how to model the relationship between protocol and release version here. If they don't, we might want to come up with a different naming scheme, or maybe some other model that eliminates the need for these names.

The (non-PRE) release protocol version list:

```
5.0 - 0x0FDB00A550010001LL
5.1 - 0x0FDB00A551040001LL
5.2 - 0x0FDB00A552000001LL
6.0 - 0x0FDB00A570010001LL
6.1 - 0x0FDB00B061060001LL
```

Reply:
I see - I guess my thought-process about mapping from version to feature to protocol version was flawed. I simplified this part a bit and simply got rid of the version-classes.